accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,258 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+join point and all of such **initialization inferred types** are syntactically
+equal, the variable is treated after the join point as having that
+**initialization inferred type**.  The promoted type chains for the variable on
+the respective paths are then intersected as usual.
+
+At a join point in the program, if an **untyped variable** has been given an
+**initialization inferred type** on more than one of the reachable paths to the
+join point and any two such **initialization inferred types** are not
+syntactically equal, it is an error.
+
+
+```dart
+void test(bool b) {
+  var x; // x is an untyped variable
+  if (b) {
+    x = 3;
+  } else {
+    x = 3.0;
+  } // Error: x has two different types on different paths
+
+  late var y; // y is a late untyped variable
+  if (b) {
+    if (b) {
+      y = 3;
+    } else {
+      y = 3.0;
+    } // Error: y has two different types on different paths
+  }
+
+ var z; // z is an untyped variable
+ if (b) {
+   z = 3; // z is promoted to int
+ } else {
+   z = 3.0; // z is promoted to double
+   return;
+ } // No error.  z is inferred as int on the only reachable path
+ z.isEven; // No error, only one reachable path
+
+ var u; // u is an untyped variable
+ if (b) {
+   u = 0 as num; // u is promoted to num
+   if (u is! int) u = 1;
+   // u has inferred type num, but is promoted to int
+ } else {
+   u = 3;
+ } // Error: u is inferred to different types on different paths
+}
+```
+
+It is an error to read an **untyped variable** when it has no **initialization
+inferred type**.
+
+
+```dart
+void test(bool b) {
+  var x; // x is an untyped variable
+  print(x); // Error, no initialization inferred type
+
+  late var y; // y is a late untyped variable
+  var f = () => print(y); // Error, y has no initialization inferred type
+
+
+  var z;
+  do {
+    // z is potentially assigned.
+    if (b) {
+      z = 42;
+    } else {
+      z = 37;
+    }
+  } while (something);
+  print(z); // Error, z has no initialization inferred type
+```
+
+As a consequence of the above it is impossible, in an error-free program, to
+observe the static type or contents of an untyped variable before it has been
+given an inferred type.  From this standpoint it may be thought of as having no
+type.  However, IDEs and tools may wish to report a static type for such a
+variable for the purposes of documentation and error reporting.  For such
+purposes, tools shall treat an untyped variable as if it were declared with a
+type equal to the upper bound of all of the **initialization inferred types**
+that it is ascribed in the method, plus `Never`.  The latter implies that a
+variable which is ascribed no **initialization inferred types** on any path
+shall be treated as having declared type `Never`.  Note that despite the
+requirement that the **initialization inferred types** be equal at join points,
+it is possible for a variable to take on different **initialization inferred
```

Comment:
We could alternatively specify it to be an error if a variable ever takes on different initialization inferred types.

Reply:
This would still allow a single type to be used, but anything with subtypes needs to be declared. Initializing to `0` and `1.0` would be invalid, not `null` (and avoid the LUB).
If you initialize with two different Widget subclasses, you need to say Widget on the variable.

It does mean that `var x; if (test) { x = e1; } else {x = e2;}` doesn't work like `var x = test ? e1 : e2;`, but there are so many ways it can not do that, and I'm not particularly fond of our LUB to begin with.

It could probably work. And it's a safe initial  behavior, we can always improve later if we make it an error now.